src/lib/keccak.ts

```diff
@@ -203,24 +380,225 @@ function round(state: Field[][], rc: Field): Field[][] {
 
 // Keccak permutation function with a constant number of rounds
 function permutation(state: Field[][], rc: Field[]): Field[][] {
-  return rc.reduce(
-    (currentState, rcValue) => round(currentState, rcValue),
-    state
+  return rc.reduce((acc, value) => round(acc, value), state);
+}
+
+// KECCAK SPONGE
+
+// Absorb padded message into a keccak state with given rate and capacity
+function absorb(
+  paddedMessage: Field[],
+  capacity: number,
+  rate: number,
+  rc: Field[]
+): Field[][] {
+  let state = getKeccakStateZeros();
+
+  // (capacity / 8) zero bytes
+  const zeros = Array(capacity / 8).fill(Field.from(0));
+
+  for (let idx = 0; idx < paddedMessage.length; idx += rate / 8) {
+    // split into blocks of rate bits
+    // for each block of rate bits in the padded message -> this is rate/8 bytes
+    const block = paddedMessage.slice(idx, idx + rate / 8);
+    // pad the block with 0s to up to 1600 bits
+    const paddedBlock = block.concat(zeros);
+    // padded with zeros each block until they are 1600 bit long
+    assert(
+      paddedBlock.length * 8 === KECCAK_STATE_LENGTH,
+      `improper Keccak block length (should be ${KECCAK_STATE_LENGTH})`
+    );
+    const blockState = getKeccakStateOfBytes(paddedBlock);
+    // xor the state with the padded block
+    const stateXor = keccakStateXor(state, blockState);
```

Comment:
> XORing the 8 zero words with the current state will create a bunch of XOR gates, which are all no-ops, so they could just be left out.

Yes, this could be an optimization, leaving out those XORs and avoid extra rows that do nothing to the inputs.

Reply:
> by making Gadgets.xor() do nothing
if this is possible, then it sounds like a good idea